### PR TITLE
fix(#1094): fix nested jar url issue

### PR DESCRIPTION
**Pull request by tschlat**

### CODE REVIEW

File: core/citrus-api/src/main/java/org/citrusframework/spi/ClasspathResourceResolver.java [hunk 1]
```diff
@@ -157,15 +157,12 @@ private static void loadFromNestedJar(ClassLoader classLoader, String path, Stri
...
         List<String> entries = new ArrayList<>();
-        try (JarInputStream jarStream = new JarInputStream(jarInputStream);) {
+        try (JarInputStream jarStream = new JarInputStream(jarInputStream)) {
             JarEntry entry;
             while ((entry = jarStream.getNextJarEntry()) != null) {
                 final String name = entry.getName().trim();
-                if (!entry.isDirectory() && filter.test(name)) {
-                    // name is FQN so it must start with package name
-                    if (name.startsWith(path)) {
+                if (!entry.isDirectory() && filter.test(name) && (name.startsWith(path))) {
```
Comment (bbortt) [inline core/citrus-api/src/main/java/org/citrusframework/spi/ClasspathResourceResolver.java:164]:
```suggestion
                if (!entry.isDirectory() && filter.test(name) && name.startsWith(path)) {
```

File: core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java [hunk 2]
```diff
@@ -8,20 +8,32 @@
...
 
 class ClassPathResourceResolverTest {
 
-    @Test
-    void loadFromFatJar() throws IOException {
+    private final ClasspathResourceResolver fixture = new ClasspathResourceResolver();
+
+
+
+
```
Comment (bbortt) [inline core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java:27]:
```suggestion

```

File: core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java [hunk 3]
```diff
@@ -30,17 +42,30 @@ void loadFromFatJar() throws IOException {
                 Path.of("META-INF/citrus/test/parser/core/xpath-data-dictionary")));
         } finally {
             Thread.currentThread().setContextClassLoader(contextClassLoader);
-        }
+        }    }
+
+    static Stream<Arguments> loadFromFatJarSpringBoot() {
```
Comment (bbortt) [inline core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java:47]: I would place the supplier method above the test, for better readability.

File: core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java [hunk 4]
```diff
@@ -30,17 +42,30 @@ void loadFromFatJar() throws IOException {
                 Path.of("META-INF/citrus/test/parser/core/xpath-data-dictionary")));
         } finally {
             Thread.currentThread().setContextClassLoader(contextClassLoader);
-        }
+        }    }
```
Comment (bbortt) [inline core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java:45]:
```suggestion
        }
    }
```

File: core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java [hunk 5]
```diff
@@ -73,9 +100,17 @@ public Enumeration<URL> getResources(String name) throws IOException {
...
+                Objects.requireNonNull(url);
+
                 URL jarResourceUrl = new URL("jar:" + url.toString().replace("\\", "/")
                     + nestedJar+ "!/META-INF/citrus/test/parser/core");
-                return Collections.enumeration(List.of(jarResourceUrl));
+
+                // "nested" is not recognized protocol and can thus not be used for creating URLS.
+                // Therefore, use a spy to fake in the "nested" protocol if needed.
+                URL urlSpy = Mockito.spy(jarResourceUrl);
+                Mockito.doReturn(jarResourceUrl.getFile().replace("file:", nestedProtocol+":")).when(urlSpy).getFile();
```
Comment (bbortt) [inline core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java:111]:
```suggestion
                doReturn(jarResourceUrl.getFile().replace("file:", nestedProtocol+":")).when(urlSpy).getFile();
```

File: core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java [hunk 6]
```diff
@@ -73,9 +100,17 @@ public Enumeration<URL> getResources(String name) throws IOException {
 
             if (name.equals("META-INF/citrus/test/parser/core/")) {
                 URL url = delegate.getResource(baseJar);
+                Objects.requireNonNull(url);
```
Comment (bbortt) [inline core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java:103]:
```suggestion
                requireNonNull(url);
```

twice, also https://github.com/citrusframework/citrus/pull/1096/files#diff-6668f866353876072e237647ff170d14eb06f78ff5e5eaea7ddbf98e84f9ba6cR135.

File: core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java [hunk 7]
```diff
@@ -8,20 +8,32 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mockito;
 
 class ClassPathResourceResolverTest {
 
-    @Test
-    void loadFromFatJar() throws IOException {
+    private final ClasspathResourceResolver fixture = new ClasspathResourceResolver();
```
Comment (bbortt) [inline core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java:23]: as far as I remember, this should be a TestNG test. I think JUnit tests won't be executed when running via Maven. TestNG also has a concept very similar to parameterized tests, they call it data providers. have a look at methods with the annotation `@Test(dataProvider="...")` for examples.

**Review by bbortt — CHANGES_REQUESTED** [review]

File: core/citrus-api/src/main/java/org/citrusframework/spi/ClasspathResourceResolver.java (hunk 1 above)
Comment (tschlat) [reply core/citrus-api/src/main/java/org/citrusframework/spi/ClasspathResourceResolver.java:164]: Quick fix 🙈.
Done.

File: core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java (hunk 3 above)
Comment (tschlat) [reply core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java]: I wouldn't but I did.

File: core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java (hunk 5 above)
Comment (tschlat) [reply core/citrus-api/src/test/java/org/citrusframework/spi/ClassPathResourceResolverTest.java]: Done

**Review by bbortt — APPROVED** [review]
perfect, many thanks!

**Pull request merged**